File: core/src/main/java/ru/pocketbyte/locolaser/config/source/SourceSet.java
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.config.source;

import ru.pocketbyte.locolaser.resource.entity.ResMap;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * @author Denis Shurygin
 */
public class SourceSet extends Source {

    private Set<Source> mSources;
    private Source mDeafaultSource;

    public SourceSet(SourceConfig sourceConfig, Set<Source> sources, Source defaultSource) {
        super(sourceConfig);
        mSources = sources;
        mDeafaultSource = defaultSource;
    }

    @Override
    public long getModifiedDate() {
        long date = 0;
        for (Source source: mSources)
            date = Math.max(date, source.getModifiedDate());
        return date;
    }

    @Override
    public ReadResult read() {
        List<MissedValue> missedValues = null;
        ResMap resMap = null;
        for (Source source : mSources) {
            Source.ReadResult result = source.read();
            if (result.missedValues != null) {
                if (missedValues == null)
                    missedValues = new ArrayList<>();
                missedValues.addAll(result.missedValues);
            }
            if (resMap != null)
                resMap = resMap.merge(result.items);
            else
                resMap = new ResMap(result.items);
        }
        return new ReadResult(resMap, missedValues);
    }

    @Override
    public void write(ResMap resMap) {
        mDeafaultSource.write(resMap);
    }

    @Override
    public void close() {
        for (Source source: mSources)
            source.close();
    }
}

File: core/src/main/java/ru/pocketbyte/locolaser/config/source/SourceSet.kt
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.config.source

import ru.pocketbyte.locolaser.config.ExtraParams
import ru.pocketbyte.locolaser.resource.entity.ResMap
import ru.pocketbyte.locolaser.resource.entity.merge
import ru.pocketbyte.locolaser.resource.formatting.FormattingType
import ru.pocketbyte.locolaser.resource.formatting.MixedFormattingType
import ru.pocketbyte.locolaser.resource.formatting.NoFormattingType
import ru.pocketbyte.locolaser.summary.FileSummary
import kotlin.math.max

/**
 * @author Denis Shurygin
 */
class SourceSet(
    private val sources: Set<Source>,
    private val default: Source
) : Source() {

    override val formattingType: FormattingType
        get() {
            val firstItemType = if (sources.isEmpty()) {
                NoFormattingType
            } else {
                sources.elementAt(0).formattingType
            }

            return if (sources.find { it.formattingType != firstItemType } != null) {
                MixedFormattingType
            } else {
                firstItemType
            }
        }

    override val modifiedDate: Long
        get() = sources.fold(0L) { acc, source ->
            max(acc, source.modifiedDate)
        }

    override fun read(locales: Set<String>?, extraParams: ExtraParams?): ResMap? {
        return sources.fold(null as? ResMap) { resMap, source ->
            resMap.merge(source.read(locales, extraParams))
        }
    }

    override fun write(resMap: ResMap, extraParams: ExtraParams?) {
        default.write(resMap, extraParams)
    }

    override fun close() {
        sources.forEach { it.close() }
    }

    override fun summaryForLocale(locale: String): FileSummary {
        var bytes: Long = 0
        val hash = StringBuilder()

        sources.forEach { source ->
            source.summaryForLocale(locale).let {
                bytes += it.bytes
                hash.append(it.hash)
            }
        }

        return FileSummary(bytes, hash.toString())
    }
}
